mozzius/expo-dynamic-app-icon
# Add auto-generated types for icon names

## Motivation

<img width="400" alt="Screenshot 2024-11-29 at 12 44 01 PM" src="https://github.com/user-attachments/assets/b30c2c15-346e-49a3-a0ab-2ae10d444c8c">

Since we already know ahead of time what the icon names will be, let's make our getters and setters strongly typed!

## What This Does

- Adds a new `DynamicAppIconRegistry` interface for defining the `IconName` type
- Adds a new custom mod in our Expo plugin to generate the union `IconName` based on our config plugin
- Updates `example` set up to use Yarn local linking for a closer-to-prod setup

## How it Works

1. User defines the different icons in their Expo `app.json` config
2. When the development server starts, the `withGenerateTypes` plugin mod runs
3. The `types.d.ts` file in the module `build` directory is updated with the strict types
4. The user can now access those types with the `IconName` import from `@mozzius/expo-dynamic-app-icon`

## How to Test

1. Rebuild the plugin (`yarn build:plugin`) and module (`yarn build`)
5. Start the development server (`yarn start`)
6. Try to change the set call in `App.tsx` L37. Confirm it's now strictly typed!

## TODO

- [ ] Re-generate the types automatically when `app.json` changes (instead of needing to restart the dev server)
